### PR TITLE
Added support for mock plugins

### PR DESCRIPTION
**Pull request by Jurisak**
In past we have no option for mock plugins enablement.
Therefore I added simple fix to do it.

Fixes: #178

### CODE REVIEW

**Comment by phracek** [comment]
@Jurisak Seems to be fine. But a question from my side. Let's say I would like to build a package by rpmbuild or fedpkg? 
Does it work properly?
Variable mock_plugin is passed into all builders, but only mock handle it?

**Comment by phracek** [comment]
What about if mock_plugin_option is missing or enter multiple times?

**Comment by thozza** [comment]
@Jurisak the idea was to introduce something like ```--builder-params``` or ```--builder-args``` which would just take string with anything the user wants. This string then would be passed to the builder object and appended to the command before executing. It is expected that the user knows which builder he uses and what arguments it can accept.

We explicitly don't want to maintain builder specific arguments with choices which can quickly go out of sync with the reality.

**Comment by Jurisak** [comment]
@phracek:
"Seems to be fine. But a question from my side. Let's say I would like to build a package by rpmbuild or fedpkg?"
    There shouldn't be problem with rpmbuild or fedpkg cause both builder methods has **kwargs argument so unknown parameters are held there ... 

"Variable mock_plugin is passed into all builders, but only mock handle it?"
Yes as you can see on line 286 in build_helper.py. 

"What about if mock_plugin_option is missing or enter multiple times?"
This one is handled by argparse. But from what I know argparse will store the latest one ... Or the most right mock_plugin_option.

**Comment by Jurisak** [comment]
@thozza Yes I told @phracek that it's very easy to make it for all builders not just for single one. But from the issue description I understood: "Do it only for mock" ... 

About the specific arguments. I wanted to prevent typos in plugin names ... So no build would be executed if there will be problem with arguments. I think it's easy to maintain those lists but ... If you prefer just letting it on users then I will remove it...

**Comment by Jurisak** [comment]
@thozza I searched for enabling plugins in rpmbuild or fedpgk but it seems that they don't have any ... Only mock has --enable-plugin argument ... What shall I do with them?

**Comment by thozza** [comment]
@Jurisak it is in our best interest to implement more general solution than just what people are asking for in the tickets, because next month there may be another user asking for ability to pass any random argument to any other builder. The idea is not to handle the plugins use case anyhow differently. Just let the user pass any additional options to the builder and think that he know what he's doing. I would not care about typos in the string, we do not care at all. If the user passes string with additional arguments, it's his job to make sure it is accepted by the builder, otherwise it is expected that the build will fail.

**Comment by Jurisak** [comment]
@thozza Yes I totally agree but as I said the first commit was only though how it could work in the meaning that i know how to make it more general ... 

So if I understood it clearly it can be used as builder specific option enabler ... Cause for fedpkg and rpmbuild there is no specific option for enabling plugins so I could use it to pass options?

**Comment by thozza** [comment]
@Jurisak 
* rename the option.
* make it to accept arbitrary string without further checking.
* pass the string to all builders and append it to the command we already run.

**Comment by Jurisak** [comment]
@thozza Ok so no support for mock --enable-plugin? 
Take it as it is and if user wants to enable plugin must do --add-option "--enable-plugin=pm_request"? 
Also I would like to allow multiple options. Can I do that or do it only for single option?

P.S.: Just thought for myself... Didn't we moved from plugins to options?

**Comment by Jurisak** [comment]
@thozza Ok so no support for mock --enable-plugin? 
Take it as it is and if user wants to enable plugin must do --add-option "--enable-plugin=pm_request"? 
Also I would like to allow multiple options. Can I do that or do it only for single option?

P.S.: Just thought for myself... Didn't we moved from plugins to options?

**Comment by Jurisak** [comment]
Guys I found problem with option parsing.
When user types:
rebase-helper 1.6.22  --enable-option "-v" 
it got:
rebase-helper: error: argument --enable-option: expected at least one argument

No matter what kind of option you will pass it. In  "--asd"  "--wtf" same result.

After speaking with ttomecek I concider to make it really specific. Cause when user runs:
rebase-helper 1.6.22  --enable-option="-v" 
All went well ... So I suggest to write into docs how to use this option and that there must be = to work it correctly ... 

How do you see it guys?

File: rebasehelper/application.py [hunk 1]
```diff
@@ -439,6 +440,9 @@ def build_packages(self):
             results_dir = os.path.join(self.results_dir, version)
             build_dict['builds_nowait'] = self.conf.builds_nowait
             build_dict['build_tasks'] = self.conf.build_tasks
+            build_dict['enable_option'] = self.conf.enable_option
+            if self.conf.buildtool in ("copr", "fedpkg"):
+                logger.warning("We are not supporting option --enable-option for builder copr or fedpkg.")
```
Comment (phracek) [inline rebasehelper/application.py]: Personally, I guess, this can be logged as debug. User can be suppressed. But it is my opinion.

File: rebasehelper/application.py [hunk 2]
```diff
@@ -708,7 +717,7 @@ def run(self):
                     Application.check_build_requires(self.spec_file)
                 # Build packages
                 try:
-                    build = self.build_packages()
+                    build = self.build_packages
```
Comment (phracek) [inline rebasehelper/application.py]: Why did you delete brackets? Is it working? I did not know, brackets are not needed? Is this python 3 specific?

File: rebasehelper/build_helper.py [hunk 3]
```diff
@@ -327,19 +329,21 @@ def build(cls, spec, sources, patches, results_dir, root=None, arch=None, **kwar
...
-            rpms = cls._build_rpm(srpm, tmp_results_dir)
+            rpms = cls._build_rpm(srpm, tmp_results_dir, **kwargs)
             # remove SRPM - side product of building RPM
             tmp_srpm = PathHelper.find_first_file(tmp_results_dir, "*.src.rpm")
             if tmp_srpm is not None:
                 os.unlink(tmp_srpm)
 
         if rpms is None:
             # We need to be inform what directory to analyze and what spec file failed
+            print(rpm_results_dir)
```
Comment (phracek) [inline rebasehelper/build_helper.py]: Remove print and log it into debug.log

File: rebasehelper/cli.py [hunk 4]
```diff
@@ -31,13 +31,20 @@
 from rebasehelper.exceptions import RebaseHelperError
 
 
+class EnableOptionFormatException(Exception):
```
Comment (thozza) [inline rebasehelper/cli.py]: Please remove this exception and use RebaseHelperError for critical exceptions.

File: rebasehelper/cli.py [hunk 5]
```diff
@@ -143,6 +150,13 @@ def add_args(self):
             help="Specify number of retries in case build fails.",
             type=int
         )
+        self.parser.add_argument(
+            "--enable-option",
```
Comment (thozza) [inline rebasehelper/cli.py]: please rename it to indicate that this is intended only for extra arguments for builders. I already provided you with some examples.

File: rebasehelper/cli.py [hunk 6]
```diff
@@ -143,6 +150,13 @@ def add_args(self):
...
+            dest="enable_option",
```
Comment (thozza) [inline rebasehelper/cli.py]: this argument should not be needed, the value should be placed into variable named the same way by default.

File: rebasehelper/cli.py [hunk 7]
```diff
@@ -164,6 +178,9 @@ def run():
         except RebaseHelperError as e:
             logger.error('\n%s', six.text_type(e))
             sys.exit(1)
+        except EnableOptionFormatException as e:
+            sys.stderr.write("{}".format(e))
+            sys.exit(1)
```
Comment (thozza) [inline rebasehelper/cli.py]: Please remove these three lines, as the new exception is not needed. Also in the future please be consistent with the way we log things. Just have a look at the lines above your addition.

File: rebasehelper/application.py (hunk 1 above)
Comment (thozza) [inline rebasehelper/application.py]: Why this is even displayed? It will be displayed every time when user uses copr or fedpkg without even using the new option. This should be completely removed.

File: rebasehelper/application.py (hunk 2 above)
Comment (thozza) [inline rebasehelper/application.py]: This is for sure a typo. the build_packages function is not called, but instead stored in the build variable. Please fix it.

File: rebasehelper/application.py [hunk 8]
```diff
@@ -407,10 +407,11 @@ def patch_sources(self, sources):
                 OutputLogger.set_patch_output('Unapplied patches:', self.rebased_patches['unapplied'])
         OutputLogger.set_patch_output('Patches:', self.rebased_patches)
 
+    @property
```
Comment (thozza) [inline rebasehelper/application.py]: This should not be a property, please make it a function again.

File: rebasehelper/application.py [hunk 9]
```diff
@@ -407,10 +407,11 @@ def patch_sources(self, sources):
...
     def build_packages(self):
         """Function calls build class for building packages"""
         if self.conf.buildtool == 'fedpkg' and not koji_builder:
-            print ('Importing module koji failed. Switching to mockbuild.')
+            print('Importing module koji failed. Switching to mockbuild.')
```
Comment (thozza) [inline rebasehelper/application.py]: This should be modified to logging call instead of using print().

File: rebasehelper/application.py (hunk 2 above)
Comment (thozza) [inline rebasehelper/application.py]: So this is because you changed the function to a property, which actually does not make any sense. Please revert it.

File: rebasehelper/build_helper.py [hunk 10]
```diff
@@ -327,19 +329,21 @@ def build(cls, spec, sources, patches, results_dir, root=None, arch=None, **kwar
         # build SRPM
         srpm, cls.logs = cls._build_srpm(spec, sources, patches, results_dir)
 
+
```
Comment (thozza) [inline rebasehelper/build_helper.py]: please remove the extra empty line

File: rebasehelper/build_helper.py [hunk 11]
```diff
@@ -283,7 +283,7 @@ class MockBuildTool(BuildToolBase):
     logs = []
 
     @classmethod
-    def _build_rpm(cls, srpm, results_dir, root=None, arch=None):
+    def _build_rpm(cls, srpm, results_dir, root=None, arch=None, enable_option=None, **kwargs):
```
Comment (thozza) [inline rebasehelper/build_helper.py]: please remove the ```**kwargs``` as it should not be needed at all.

File: rebasehelper/build_helper.py [hunk 12]
```diff
@@ -327,19 +329,21 @@ def build(cls, spec, sources, patches, results_dir, root=None, arch=None, **kwar
...
         srpm, cls.logs = cls._build_srpm(spec, sources, patches, results_dir)
 
+
         # build RPMs
         rpm_results_dir = os.path.join(results_dir, "RPM")
         with MockTemporaryEnvironment(sources, patches, spec, rpm_results_dir) as tmp_env:
             env = tmp_env.env()
             tmp_results_dir = env.get(MockTemporaryEnvironment.TEMPDIR_RESULTS)
-            rpms = cls._build_rpm(srpm, tmp_results_dir)
+            rpms = cls._build_rpm(srpm, tmp_results_dir, **kwargs)
```
Comment (thozza) [inline rebasehelper/build_helper.py]: Why is there ```**kwargs```?

File: rebasehelper/build_helper.py [hunk 13]
```diff
@@ -371,7 +375,7 @@ class RpmbuildBuildTool(BuildToolBase):
     logs = []
 
     @classmethod
-    def _build_rpm(cls, srpm, workdir, results_dir):
+    def _build_rpm(cls, srpm, workdir, results_dir, enable_option=None, **kwargs):
```
Comment (thozza) [inline rebasehelper/build_helper.py]: please remove the ```**kwargs``` as it should not be needed at all.

File: rebasehelper/build_helper.py [hunk 14]
```diff
@@ -425,7 +431,7 @@ def build(cls, spec, sources, patches, results_dir, **kwargs):
             env = tmp_env.env()
             tmp_dir = tmp_env.path()
             tmp_results_dir = env.get(RpmbuildTemporaryEnvironment.TEMPDIR_RESULTS)
-            rpms = cls._build_rpm(srpm, tmp_dir, tmp_results_dir)
+            rpms = cls._build_rpm(srpm, tmp_dir, tmp_results_dir, **kwargs)
```
Comment (thozza) [inline rebasehelper/build_helper.py]: Why is there ```**kwargs```?

File: rebasehelper/build_helper.py [hunk 15]
```diff
@@ -386,6 +390,8 @@ def _build_rpm(cls, srpm, workdir, results_dir):
         output = os.path.join(results_dir, "build.log")
 
         cmd = [cls.CMD, '--rebuild', srpm]
+        if enable_option is not None:
+            cmd.extend(filter(None, enable_option.split(" ")))
```
Comment (thozza) [inline rebasehelper/build_helper.py]: The same code is also in Rpmbuild build tool class. Maybe it would be a good idea to extract the actual code to the BuildToolBase class and call it from there. Or create new common base class for local builders, since these are the only ones for which it makes sense.

**Comment by thozza** [comment]
@Jurisak please rebase and merge the commits, so that there are only commits with the actual changes which are currently relevant.

File: rebasehelper/application.py (hunk 9 above)
Comment (Jurisak) [inline rebasehelper/application.py]: Agree I was only disturbed by the whitespace.

File: rebasehelper/application.py (hunk 8 above)
Comment (Jurisak) [inline rebasehelper/application.py]: I didn't make it intentionally as a property. I assume that I hit some combination of keys and my IDE did this behind my back. I'm very sorry for that. It's my first case...

File: rebasehelper/application.py (hunk 1 above)
Comment (Jurisak) [inline rebasehelper/application.py]: Well i think it's better to use warning cause it should inform you that this is not working on other builders and yes there is missing and with checking enable_option ... 

For me it's good. User forgets that enable_option cant be used for copr and fedpkg and he should be informed. If you want it through debug. I can do it through debug...

File: rebasehelper/application.py (hunk 2 above)
Comment (Jurisak) [inline rebasehelper/application.py]: It's done by my IDE. Sorry for that. I'll revert it.

File: rebasehelper/build_helper.py (hunk 3 above)
Comment (Jurisak) [inline rebasehelper/build_helper.py]: This shouldn't be here at all...

File: rebasehelper/cli.py (hunk 5 above)
Comment (Jurisak) [inline rebasehelper/cli.py]: --enable-extra-builder-options ? or --enable-builder-options ? Or you have any other suggestions?

File: rebasehelper/cli.py (hunk 6 above)
Comment (Jurisak) [inline rebasehelper/cli.py]: Correct

File: rebasehelper/cli.py (hunk 4 above)
Comment (Jurisak) [inline rebasehelper/cli.py]: Ok.

File: rebasehelper/cli.py (hunk 7 above)
Comment (Jurisak) [inline rebasehelper/cli.py]: Well it's kinda is ... I wanted to stop using extra builder option in wrong format so I raised RebaseHelperError and I end up with no message at all. :D After some investigation I found that we are creating logger in Application.__init__.py so EVERY exception in CLI will not use logger cause it's created after argument parsing. 

I wanted to discuss with you this problem but you were not responding.

File: rebasehelper/build_helper.py (hunk 10 above)
Comment (Jurisak) [inline rebasehelper/build_helper.py]: Yep.

File: rebasehelper/build_helper.py (hunk 11 above)
Comment (Jurisak) [inline rebasehelper/build_helper.py]: Well my though was this. Method build is called with dictionary which represents some of method parameters. My though was that in future you will forget to add enable_option into this dict and you will end up with error. 

I can remove it and check it before the call but my second though was that in future we may use **kwargs in _build_rpm method.

File: rebasehelper/build_helper.py (hunk 12 above)
Comment (Jurisak) [inline rebasehelper/build_helper.py]: In **kwargs is dict which contains enable_option parameter. So that why.

File: rebasehelper/build_helper.py (hunk 13 above)
Comment (Jurisak) [inline rebasehelper/build_helper.py]: Well my though was this. Method build is called with dictionary which represents some of method parameters. My though was that in future you will forget to add enable_option into this dict and you will end up with error.

I can remove it and check it before the call but my second though was that in future we may use **kwargs in _build_rpm method.

File: rebasehelper/build_helper.py (hunk 14 above)
Comment (Jurisak) [inline rebasehelper/build_helper.py]: In **kwargs is dict which contains enable_option parameter. So that why.

File: rebasehelper/build_helper.py (hunk 15 above)
Comment (Jurisak) [inline rebasehelper/build_helper.py]: I will think about it.

**Comment by Jurisak** [comment]
@thozza Could you please look at my comments. I'm not 100% sure about some. Thanks :)

File: rebasehelper/application.py (hunk 1 above)
Comment (thozza) [inline rebasehelper/application.py]: When user uses copr/fedpkg it does not make sense to inform them about some option they didn't even specified on command line.

The only situation when it would make sense it when they explicitly use copr/fedpkg with the option.

File: rebasehelper/build_helper.py (hunk 11 above)
Comment (thozza) [inline rebasehelper/build_helper.py]: You won't because the option has default value. Getting error which we can fix is better than masking it. For new code explicit parameters are preferred, unless there is strong reason to take arbitrary kwargs. If there will be such situation in the future, we will add it then, now it is just dead code.

File: rebasehelper/build_helper.py (hunk 12 above)
Comment (thozza) [inline rebasehelper/build_helper.py]: Then  please pass it explicitly instead of the whole dictionary, it is more readable.

File: rebasehelper/build_helper.py (hunk 13 above)
Comment (thozza) [inline rebasehelper/build_helper.py]: The same as above...

File: rebasehelper/build_helper.py (hunk 14 above)
Comment (thozza) [inline rebasehelper/build_helper.py]: Same as above...

File: rebasehelper/cli.py (hunk 7 above)
Comment (thozza) [inline rebasehelper/cli.py]: I that case please think about moving the check and raising the exception after the parsing of arguments. Also the logging is little bit reworked in another PR which we will be merging shortly and it may help you with the problem.

File: rebasehelper/cli.py (hunk 5 above)
Comment (thozza) [inline rebasehelper/cli.py]: @Jurisak So my initial ideas were ```--builder-params``` or ```--builder-args```. IMHO there is no need to use "enable" in the option name.

File: rebasehelper/build_helper.py (hunk 12 above)
Comment (Jurisak) [inline rebasehelper/build_helper.py]: OK I admit that it is really better to have it explicit than implicit ... Sorry for that.

File: rebasehelper/application.py (hunk 2 above)
Comment (Jurisak) [inline rebasehelper/application.py]: Well this one is really strange. In my code and in my local commits I have it without property? Does anybody know how the hell the property came from?

File: rebasehelper/cli.py (hunk 5 above)
Comment (Jurisak) [inline rebasehelper/cli.py]: Ok. I agree with the --builder at the beginning but I would rather use `--builder-option` cause it logically connects with the builder option rather than with params or args ... How do you see it?

File: rebasehelper/build_helper.py [hunk 16]
```diff
@@ -272,6 +272,14 @@ def _build_srpm(cls, spec, sources, patches, results_dir):
 
         return srpm, logs
 
+    @classmethod
```
Comment (thozza) [inline rebasehelper/build_helper.py]: this could be static

File: rebasehelper/build_helper.py [hunk 17]
```diff
@@ -272,6 +272,14 @@ def _build_srpm(cls, spec, sources, patches, results_dir):
...
+    def get_builder_options(cls, **kwargs):
+        if 'builder_options' not in kwargs:
+            return None
+        if kwargs['builder_options'] is not None:
```
Comment (thozza) [inline rebasehelper/build_helper.py]: This could be simplified to:
```python
builder_options = kwargs.get('builder_options', None)
if builder_options is not None:
    ...
```

File: rebasehelper/cli.py [hunk 18]
```diff
@@ -163,8 +174,8 @@ class CliHelper(object):
 
     @staticmethod
     def run():
+        debug_log_file = None
         try:
-            debug_log_file = None
```
Comment (thozza) [inline rebasehelper/cli.py]: why did you do this change?

File: rebasehelper/build_helper.py (hunk 16 above)
Comment (Jurisak) [inline rebasehelper/build_helper.py]: Yes it could but I make it as a classmethod cause rest of the method are classmethods as well...

File: rebasehelper/build_helper.py (hunk 17 above)
Comment (Jurisak) [inline rebasehelper/build_helper.py]: I never know about this. I like it :).

File: rebasehelper/cli.py (hunk 18 above)
Comment (Jurisak) [inline rebasehelper/cli.py]: I think catching exception is only for code which can raise an exception. Also you are using `debug_log_file` in one of the exceptions and my IDE marked it that could be referenced before assignment. I know that it's on the first line BUT still I think try: catch: clause should contain code which could raise an Exception and `debug_log_file = None` will not ..

File: rebasehelper/cli.py [hunk 19]
```diff
@@ -163,8 +174,8 @@ class CliHelper(object):
...
             # be verbose until debug_log_file is created
             handler = LoggerHelper.add_stream_handler(logger, logging.DEBUG)
             cli = CLI()
```
Comment (Jurisak) [inline rebasehelper/cli.py]: I think it must be CLI(sys.argv[1:]) ... I fixed in next commit.

File: rebasehelper/cli.py (hunk 18 above)
Comment (nforro) [inline rebasehelper/cli.py]: I think this change is ok, but it is outside the scope of this pull request.

File: rebasehelper/cli.py (hunk 19 above)
Comment (nforro) [inline rebasehelper/cli.py]: This is intentional, since `sys.argv[1:]` is what `ArgumentParser.parse_args` uses if not provided with any arguments.

File: rebasehelper/cli.py (hunk 18 above)
Comment (Jurisak) [inline rebasehelper/cli.py]: Agree there is more refactorization ... I'll try to avoid this next time.

File: rebasehelper/cli.py (hunk 19 above)
Comment (Jurisak) [inline rebasehelper/cli.py]: Ok. According to this i need to do an update.

**Comment by thozza** [comment]
closing in favor of PR #216 so that we can do a new release

**Comment by thozza** [comment]
@Jurisak thank you for the contribution.
